### DISCUSSION TITLE
Ability to autoselect methods only when result type is equal to the desired.

### DISCUSSION THREAD

**Discussion by mihilion** [mapstruct/mapstruct#2816] [Ideas]
Can you please add possibility to define mapping methods, which will be auto selected only when target type is strictly equal to the desired target type, and not when it is assignable to that type?

Example, think about two DTOs, one extends another, and two mapping methods:
```
public class UserSimpleDto { ... }
public class UserDto extends UserSimpleDto { ... }

public abstract UserSimpleDto userToUserSimpleDto(User user);
public abstract UserDto userToUserDto(User user);
```
Then we have object with field `user` of type `UserSimpleDto` and mapping method for it
```
public class ContractSimpleDto {
  private UserSimpleDto user;
}

public abstract ContractSimpleDto contractToContractSimpleDto(Contract contract);
```
Mapstruct now raises ambiguous method exception for field `user`, because both `UserSimpleDto` and `UserDto` are assignable to `UserSimpleDto`. This behaviour is not what I want. I would like to be able to (either globally, or at least directly on mapping methods) tell mapstruct to strictly select method with target type equal to `UserSimpleDto`, and ignore mapping to `UserDto` during automatic method discovery.

I know I can use `@Mapping(target = 'user', resultType = UserSimpleDto.class)` on contract mapping method, but I have lots of objects with many fields of type `UserSimpleDto`, so there are a lot of such occurences in the code. It would be much simpler to just make methods for mapping `User` only discoverable when the resultType strictly matches.

My workaround for this problem looks like this:
```
@QualifedByName("notAutoSelected")
public abstract UserSimpleDto userToUserSimpleDto(User user);

public UserSimpleDto userToUserSimpleDto(User user, @TargetType Class<UserSimpleDto> clazz) {
    return userToUserSimpleDto(user);
}

@QualifidByName("notAutoSelected")
public abstract UserDto userToUserDto(User user);

public UserDto userToUserDto(User user, @TargetType Class<UserDto> clazz) {
    return userToUserDto(user);
}
```
But this is a lot of boilerplate code. Using a different mapping strategy for the whole mapper, and/or for individual methods would make everything much simpler.

**Comment by Zegveld** [comment]
Just curious, but why not create more Mapper classes?
You could then use the `uses` directive to select which mapper is used where.
For example:
```
UserMapper
SimpleUserMapper
ContractMapper --> uses UserMapper
SimpleContractMapper --> uses SimpleUserMapper
```

**Reply by mihilion** [reply]
Sticking to the example, I have cases like this, where I would have to add both `UserMapper` and `SimpleUserMapper` to `uses`, and the problem would remain:
```
public class ContractDto {
    private UserDto owner;
    private UserSimpleDto modifiedBy;
}
```

By the way, I already have over 100 Mapper classes for different objects, and I use `uses` all the time, but each Mapper class handles one object at all detail levels. And since I have not 2 but 4 levels of details for most of the objects (`ContractDto extends ContractSimpleDto extends ContractBaseDto extends ContractIdDto`), splitting each Mapper to 4 classes would produce ~300 more classes, and still not solve the problem. Moreover, navigating through over 100 Mappers is already challenging - people make mistakes in this maze and often use wrong mappers, also when type of field changes all mappers need to be reviesed to adjust the mapper which should be used. All this mess could be eliminated with something as simple as setting one property on Mapper level.

The default Mapstruct behaviour can even result in producing Mappers which do not compile! Please take a look at following example:
```
public class ContractDto {
    private List<UserSimpleDto> someUsers;
}

public abstract ContractDto contractToContractDto(Contract contract);
public abstract List<UserDto> userListToUserDtoList(List<User> list);
```
Now mapstruct will decide, that `userListToUserDtoList` is perfect match to map `someUsers`, but although `UserDto` is assignable to `UserSimpleDto`, `List<UserDto>` is NOT assignable to `List<UserSimpleDto>`, so code generated by Mapstruct does not compile.

If I do not define method for mapping list of users, Mapstruct generates correct one. However if I define two methods to map list of users (one to `List<UserDto>` and another one to `List<UserSimpleDto>`), Mapstruct raises ambiguous methods found exception, although one of these methods does not work here. To resolve it, I need to qualifiy both methods, and explicitly choose one in `@Mapping` annotation. With strict result type matching, this problem would not exist. I hope that this additional example with auto-selecting method for mapping List (which is actually also a valid problem for me) will allow you to see things from my point of view. I really like Mapstruct, and I heavily use it. If I am using it wrong, I am open for discussion and and hints on how to use it better, but I really think that adding the suggested option would make life much easier (and for collections it should actually be the only valid strategy, unless someone defines a list as `List<? extends UserSimpleDto>`)
